### ISSUE TITLE
Work out whether any changes are needed for positional/keyword args in Ruby 3.0

### ISSUE THREAD

**Issue by floehopper** [freerange/mocha#446]
See: https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/

**Comment by wasabigeek** [comment]
I'm having trouble matching keyword arguments precisely in Ruby 3. The problem seems to be that the `expected_parameters` look the same for both these cases:
- `.with("example", { foo => "bar" })`  => expected parameters are `[“example”, { foo => “bar” }]`
- `.with("example", foo => "bar")` (or `.with("example", **{ foo => "bar"})`) => expected parameters are also `[“example”, { foo => “bar” }]`

But they aren't the same in Ruby 3. For example:

```ruby
require "minitest/autorun"
require 'mocha/minitest'
class Example
  def splatted_foo(*args, **kwargs)
    modified_kwargs = kwargs.merge({ example: :text })
    bar(*args, **modified_kwargs)
  end

  def nonsplatted_foo(*args, **kwargs)
    modified_kwargs = kwargs.merge({ example: :text })
    bar(*args, modified_kwargs)
  end

  def bar(*args, hello:, example:)
  end
end

class Test < Minitest::Test
  def test_wrapper_passes_arguments
    example = Example.new
    args = [1]
    kwargs = {hello: :world}
    expected_kwargs = {hello: :world, example: :text}
    example.expects(:bar).with(1, expected_kwargs) # passes
    example.splatted_foo(*args, **kwargs)

    example.expects(:bar).with(1, expected_kwargs) # passes
    example.nonsplatted_foo(*args, **kwargs) # however, this same code would raise an ArgumentError if run in console
  end
end
```
Above was tested with ruby 3.0.3, minitest (5.15.0), mocha (1.13.0).

**Comment by floehopper** [comment]
@wasabigeek Thanks for reporting this. I'm on holiday at the moment, so I might not be able to look at this properly for a while. Just to make sure I understand, can you confirm that the main issue is that the stubbed version of `Example#foo` doesn't behave the same way as the real version of the method, because the stubbed version does not raise an `ArgumentError` in cases where the real method would do so...?

**Comment by wasabigeek** [comment]
> the main issue is that the stubbed version of Example#foo doesn't behave the same way as the real version of the method, because the stubbed version does not raise an ArgumentError in cases where the real method would do so...?

More precisely, I'm trying to set a `with` expectation to specifically check that _keyword arguments_ are being passed in, but Mocha currently treats last parameter hashes and keyword arguments the same, causing the test to pass when it actually would fail when executing the code.

Sorry if I'm not clear, finding it hard to explain this 🙇 Perhaps another example:
```ruby
require "minitest/autorun"
require 'mocha/minitest'
class Example
  def foo(arg, kwarg:)
  end
end

class Test < Minitest::Test
  def test_arguments
    example = Example.new
    arg = "test1"
    hash = { kwarg: "test2" }
    example.expects(:foo).with(arg, **hash).twice

    # This is the correct form, splatting the hash as kwargs.
    # This passes.
    example.foo(arg, **hash)

    # This is an incorrect form, as in Ruby 3 the hash is not interpreted as kwargs (see example error later).
    # However, this also passes! So this code might have made it into production.
    # I would have expected this to have been caught by the `with` expectation.
    example.foo(arg, hash)

    # To illustrate the actual error
    example.unstub(:foo)
    example.foo(arg, hash) # ArgumentError: wrong number of arguments (given 2, expected 1; required keyword: kwarg)
  end
end
```

TIL `Method#parameters` 👍 I think that's only half of the puzzle though, because while trying to patch mocha I found that all the args and kwargs get lumped into a single array. For example:
```ruby
class Invocation
  def initialize(mock, method_name, *arguments, &block)
    # both the following would give the same `arguments`, when in the first, it could have been a positional hash argument
    # Invocation.new(mock, method_name, "arg1", { kwarg: "test" })
    # Invocation.new(mock, method_name, "arg1", kwarg: "test")
    ...
  end
end
```
I'm not sure if there's a good way to distinguish this - we could try converting references for `*arguments` into `*args, **kwargs` but I think this meets with some trouble at `method_missing`:
```ruby

def example(_arg, options = {})
  p options
end

def example_method_missing(*args, **kwargs)
  p [args, kwargs]
end

# This matches the signature exactly, and prints the correct options.
example("", {test: 2}) # {:test=>2}
# if we use method missing with this style, it interprets the hash as a positional argument, as expected.
example_method_missing("", {test: 2}) # [["", {:test=>2}], {}]

# However, Ruby 3 still allows this, where the trailing kwargs gets treated as a positional param:
example("", test: 2) # {:test=>2}
# In this scenario, the trailing kwargs get interpreted as kwargs, instead of a positional param, which is different from above:
example_method_missing("", test: 2) # [[""], {:test=>2}]
```

**Comment by wasabigeek** [comment]
@floehopper wondering if you've any thoughts on how to address this? Would love to help where possible.

Thought: we could replace `*args` with the `*args, **kwargs` syntax, and add some abstraction that (depending on preference) either lets users access the args and kwargs separately, or access both args and kwargs as an array to mimic the current behaviour. This would give users an upgrade path of sorts. What do you think?

**Comment by floehopper** [comment]
@wasabigeek Hi. Thanks for your patience on this. I'm afraid I haven't got any time right now, but it might be worth you having a look at the changes in #475 to see if they help. Let me know and I'll try to find some time to think about this more carefully soon! 🤞

**Comment by floehopper** [comment]
Actually, this one which I've been giving some recent work on might be relevant too: #149.

**Comment by wasabigeek** [comment]
The particular concern I have is that keyword matching is imprecise and may cause a false negative (test passes when it fails in real life). From what I recall, there are two parts to the problem:
1. determining the arity / parameter types of the mocked method
2. matching `with` params to the corresponding args in the mocked method

For (1), I think `Method#parameters` will distinguish between what's a keyword arg and what's not, so that sounds good 👍 (at least, that's what I am assuming will be used in #149).
For (2), the problem seems to be that all `with` args are lumped into a single array (`@expected_parameters`). #475 assumes the [last arg](https://github.com/freerange/mocha/pull/475/files#diff-5440fe2ba7746f3290444b3d79be4fa757b7c4b6d489cf6989a66ad478795a51R7-R8) is assumed to be a keyword arg if it's a hash, when this is not always true in Ruby 3. For example, given `def example(foo, bar:);end`, run the following in irb:
```ruby
irb(main):004:0> example('a', bar: 'b')
=> nil
irb(main):005:0> example('a', { bar: 'b' })
(irb):3:in `example': wrong number of arguments (given 2, expected 1; required keyword: bar) (ArgumentError)
```
Above, if mocking `.with('a', { bar: 'b'})` the test would pass, when it's actually not a valid set of arguments.

**Comment by wasabigeek** [comment]
I think this is a related issue in RSpec https://github.com/rspec/rspec-support/issues/522#issuecomment-999772629 with an attempted fix https://github.com/rspec/rspec-support/pull/537. Not super familiar with the RSpec internals though.

**Comment by wasabigeek** [comment]
@floehopper I've attempted a proof of concept for checking keyword args more strictly via `ruby2_keywords` in https://github.com/freerange/mocha/pull/534. It's not the most graceful but it seems to work, would love to hear your thoughts.

**Comment by floehopper** [comment]
@wasabigeek Thanks so much for continuing to work on this. I've just had some time to read this issue through more carefully and I think I'm slowly getting my head around it. I'll have a more careful look at #534 now. Can I just double-check the latter is only intended to address point (2) in [your earlier comment](https://github.com/freerange/mocha/issues/446#issuecomment-1221215369), i.e. keyword matching using `Expectation#with` and not raising `ArgumentError` when the method signature can't handle the supplied parameters?

**Comment by wasabigeek** [comment]
Yes, it only addresses (2)! At first, I thought (1) might be a prerequisite, but it seems it's not.

**Comment by wasabigeek** [comment]
@floehopper, I think explicit kwarg matching can work, but there is some ambiguity in the current API when hash matchers are involved e.g. is `.with(has_value(“hello”))` matching a last positional Hash, or a kwarg? I think we can possibly:
- try to allow hash matchers to check for both last positional Hash _and_ kwargs, passing if either scenario matches (which would be ambiguous but shouldn't break existing test suites, and maybe we can phase this out later in favour of explicitness)
- introduce a way to explicitly declare a kwarg matcher, e.g.:
    - `.with(arg, kwargs(has_value(“hello”)))`
    - `.with(arg, kwargs: has_value(“hello”))`

What do you think?

**Comment by wasabigeek** [comment]
Tracking some todos now that Ruby 1.9 support is dropped:
- [ ] Add keyword argument tests for existing behaviour https://github.com/freerange/mocha/pull/551
- [ ] Add ruby2_keywords shim
- [ ] Add configuration and strict argument checking feature https://github.com/freerange/mocha/pull/554

**Comment by floehopper** [comment]
> Tracking some todos now that Ruby 1.9 support is dropped:
> 
> * [ ]  [Add acceptance tests for keyword arguments #551](https://github.com/freerange/mocha/pull/551) for existing behaviour
> * [ ]  [Add ruby2_keywords shim](https://github.com/freerange/mocha/pull/559)
> * [ ]  Add configuration and strict argument checking feature [Support strict keyword argument matching #554](https://github.com/freerange/mocha/pull/554)
> 
> The above follows the template laid out in this spike: #544

@wasabigeek This makes sense to me, although I'm starting to think I might like to combine all these changes into a single PR at least when I come to merge all the changes. I'll give it some more thought...

**Comment by wasabigeek** [comment]
I'm happy to combine it! That was a thought that went through my head. I'll work on that this week.
